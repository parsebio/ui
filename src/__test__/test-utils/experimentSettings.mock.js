import _ from 'lodash';

import initialState from 'redux/reducers/experimentSettings/initialState';
import { analysisTools } from 'utils/constants';

const sampleifiedConfig = (sampleIds, configToReplicate) => {
  const result = sampleIds.reduce(
    (acum, sampleId) => {
      // eslint-disable-next-line no-param-reassign
      acum[sampleId] = configToReplicate;
      return acum;
    },
    {},
  );

  return result;
};

const generateProcessingConfigMock = (sampleIds) => ({
  classifier: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      prefiltered: false,
      filterSettings: {
        FDR: 0.1,
      },
    }),
  },
  cellSizeDistribution: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        minCellSize: 10800,
        binStep: 200,
      },
    }),
  },
  mitochondrialContent: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        method: 'absoluteThreshold',
        methodSettings: {
          absoluteThreshold: {
            maxFraction: 0.1,
            binStep: 200,
          },
        },
      },
    }),
  },
  numGenesVsNumUmis: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        regressionType: 'gam',
        regressionTypeSettings: {
          gam: {
            'p.level': 0.00009,
          },
        },
      },
    }),
  },
  doubletScores: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        probabilityThreshold: 0.2,
        binStep: 0.05,
      },
    }),
  },
  dataIntegration: {
    enabled: true,
    analysisTool: analysisTools.SEURAT,
    dataIntegration: {
      method: 'harmony',
      methodSettings: {
        harmony: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      useSaved: false,
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
});

const generateExperimentSettingsMock = (sampleIds) => {
  const mockedProcessingConfig = {
    ...initialState.processing,
    ...generateProcessingConfigMock(sampleIds),
  };

  return {
    ...initialState,
    processing: mockedProcessingConfig,
    originalProcessing: _.cloneDeep(mockedProcessingConfig),
    info: {
      ...initialState.info,
      pipelineVersion: 1,
    },
  };
};

export default generateExperimentSettingsMock;
