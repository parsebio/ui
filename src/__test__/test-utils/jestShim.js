/**
 * This file is required for jest to work with '@cfaester/enzyme-adapter-react-18'
 * module.
 */
import {
  TextDecoder, TextEncoder,
} from 'util';

global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
