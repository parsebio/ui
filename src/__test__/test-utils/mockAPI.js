import _ from 'lodash';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';
import { mockSecondaryAnalyses } from '__test__/data/secondaryAnalyses/secondary_analyses';
import mockSecondaryAnalysisStatusDefault from '__test__/data/secondaryAnalyses/secondary_analysis_status_default.json';
import fake from '__test__/test-utils/constants';
import mockAnalysisFiles from '__test__/data/secondaryAnalyses/secondary_analysis_files';
import downloadFromS3 from 'utils/work/downloadFromS3';

import {
  responseData,
} from '__test__/test-utils/mockData';
import { analysisTools } from 'const';

const cellSetsData = require('__test__/data/cell_sets.json');

const setupDownloadCellSetsFromS3Mock = (customCellSets = cellSetsData) => {
  downloadFromS3.mockImplementation((resource, signedUrl) => {
    // eslint-disable-next-line global-require
    if (signedUrl === 'mock-cellsets-signed-url') {
      return Promise.resolve(customCellSets);
    }
    return Promise.reject(new Error('Invalid signed URL'));
  });
};

const promiseResponse = (response, options = {}) => (
  Promise.resolve(new Response(response, options)));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (mockedResults) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }
  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (projectId, analysisTool = analysisTools.SEURAT) => ({
  [`experiments/${projectId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === projectId)),
  ),
  [`/v2/experiments/${projectId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === projectId)),
  ),
  [`experiments/${projectId}/processingConfig$`]: () => {
    const processingConfigDataClone = _.cloneDeep(processingConfigData);
    processingConfigDataClone.processingConfig.dataIntegration.analysisTool = analysisTool;

    return promiseResponse(
      JSON.stringify(processingConfigDataClone),
    );
  },
  [`experiments/${projectId}/cellSets$`]: () => {
    setupDownloadCellSetsFromS3Mock();
    return promiseResponse(
      JSON.stringify('mock-cellsets-signed-url'),
    );
  },
  [`experiments/${projectId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${projectId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${projectId}`]: () => statusResponse(200, 'OK'),
  '/v2/secondaryAnalysis$': () => {
    const mockedMainSecondary = _.cloneDeep(mockSecondaryAnalyses[1]);
    mockedMainSecondary.id = projectId;

    return promiseResponse(
      JSON.stringify([
        mockedMainSecondary,
        ...mockSecondaryAnalyses,
      ]),
    );
  },
  [`/v2/secondaryAnalysis/${projectId}/executionStatus`]: () => promiseResponse(
    JSON.stringify(mockSecondaryAnalysisStatusDefault),
  ),
  [`/v2/secondaryAnalysis/${projectId}/files`]: (req) => {
    if (req.method === 'GET') {
      return promiseResponse(
        JSON.stringify(mockAnalysisFiles),
      );
    }

    if (req.method === 'POST') {
      return promiseResponse(
        JSON.stringify({ fileId: 'mockFileId' }),
      );
    }
  },
  [`/v2/projects/${projectId}/upload/.*/part/\\d+/signedUrl$`]: (req) => {
    const partNumber = req.url.match(/part\/(\d+)\/signedUrl/)[1];
    return promiseResponse(JSON.stringify(`mockSignedUrl${partNumber}`));
  },
  '/v2/cliUpload/token$': () => promiseResponse(JSON.stringify('token')),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
  setupDownloadCellSetsFromS3Mock,
};
