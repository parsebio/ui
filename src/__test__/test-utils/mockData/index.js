import _ from 'lodash';
import { analysisTools } from 'utils/constants';
import generateMockSamples from './generateMockSamples';
import generateMockExperiments from './generateMockExperiments';
import generateMockProcessingConfig from './generateMockProcessingConfig';

// This file is a work in progress to generate mock data for the endpoints of the test.
// The generated data should be usable as response data to be sent via the mocked API in mockAPI.js
// The module should also expose the generated data that can be imported and used easily in tests

const responseData = {};

responseData.experiments = generateMockExperiments(2);

responseData.samples = [generateMockSamples(
  responseData.experiments[0],
  3,
)];

responseData.processingConfigSeurat = generateMockProcessingConfig(analysisTools.SEURAT, 3);
responseData.processingConfigScanpy = generateMockProcessingConfig(analysisTools.SCANPY, 3);

// Add samples to first experiment
const samples = responseData.samples[0];
const sampleIds = _.map(samples, 'id');
responseData.experiments[0].samplesOrder = sampleIds;

const { experiments, processingConfigSeurat, processingConfigScanpy } = responseData;

export {
  experiments,
  samples,
  processingConfigSeurat,
  processingConfigScanpy,
  responseData,
};
