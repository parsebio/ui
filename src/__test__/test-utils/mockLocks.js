const setupNavigatorLocks = () => {
  const mockLockRequest = jest.fn((lock, func) => func());
  global.navigator.locks = {
    request: mockLockRequest,
  };

  return mockLockRequest;
};

const teardownNavigatorLocks = () => {
  delete global.navigator.locks;
};

export { setupNavigatorLocks, teardownNavigatorLocks };
