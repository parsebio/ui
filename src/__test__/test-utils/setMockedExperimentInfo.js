import { updateExperimentInfo } from 'redux/actions/experimentSettings';

const setMockedExperimentInfo = async (experimentId, store) => {
  await store.dispatch(
    updateExperimentInfo({
      experimentId,
      experimentName: 'mockedName',
      sampleIds: ['1', '2'],
      pipelineVersion: 'v2',
      accessRole: 'owner',
    }),
  );
};

export default setMockedExperimentInfo;
