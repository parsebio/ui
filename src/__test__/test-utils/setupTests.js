import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
}));

jest.mock('utils/work/downloadFromS3');

// The updated versions of react - dnd seem to not have a compatible version
// of the testing libraries anymore, so trying anything with the old way of setting up tests
// (with wrapWithTestBackend from react-dnd-test-utils or similar options) breaks.

jest.mock('react-dnd', () => {
  const mockBackend = { /* add methods as needed for your tests */ };
  const mockManager = {
    getBackend: jest.fn(() => mockBackend),
    getMonitor: jest.fn(() => ({
      subscribeToStateChange: jest.fn(),
      subscribeToOffsetChange: jest.fn(),
      isDragging: jest.fn(() => false),
      getItemType: jest.fn(),
      getItem: jest.fn(),
      canDragSource: jest.fn(),
      canDropOnTarget: jest.fn(),
      getDropResult: jest.fn(),
      didDrop: jest.fn(),
      getInitialClientOffset: jest.fn(),
      getInitialSourceClientOffset: jest.fn(),
      getSourceClientOffset: jest.fn(),
      getClientOffset: jest.fn(),
      getDifferenceFromInitialOffset: jest.fn(),
      isOverTarget: jest.fn(() => false),
      getTargetIds: jest.fn(),
    })),
    getActions: jest.fn(() => ({
      beginDrag: jest.fn(),
      publishDragSource: jest.fn(),
      hover: jest.fn(),
      drop: jest.fn(),
      endDrag: jest.fn(),
    })),
    getRegistry: jest.fn(() => ({
      addSource: jest.fn(),
      addTarget: jest.fn(),
      removeSource: jest.fn(),
      removeTarget: jest.fn(),
    })),
  };
  return {
    useDrag: () => [{ isDragging: false, handlerId: 'mock', draggedItem: null }, jest.fn(), jest.fn()],
    useDrop: () => [{ isOver: false, canDrop: false, handlerId: 'mock' }, jest.fn()],
    useDragDropManager: () => mockManager,
    DndProvider: ({ children }) => children,
  };
});

jest.mock('react-dnd-multi-backend', () => ({
  TouchTransition: {},
  HTML5toTouch: {},
  DndProvider: ({ children }) => children,
  createTransition: jest.fn(),
  default: {},
}));

jest.mock('rdndmb-html5-to-touch', () => ({
  default: {},
  HTML5toTouch: {},
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});

// setImmediate is not available in JSDOM, only on node
// https://github.com/prisma/prisma/issues/8558#issuecomment-1006100001
global.setImmediate = global.setImmediate || ((fn, ...args) => global.setTimeout(fn, 0, ...args));
