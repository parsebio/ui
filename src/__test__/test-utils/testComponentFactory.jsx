/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import _ from 'lodash';

const createTestComponentFactory = (Component, defaultProps = {}) => (customProps = {}) => {
  // Merge is given an empty object so that it always
  // return a new object without mutating defaultProps
  const props = _.merge(
    {},
    defaultProps,
    customProps,
  );

  return <Component {...props} />;
};

export default createTestComponentFactory;
