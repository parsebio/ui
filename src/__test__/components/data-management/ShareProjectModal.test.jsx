import React from 'react';
import _ from 'lodash';
import {
  screen, render, fireEvent, waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { makeStore } from 'redux/store';

import { act } from 'react-dom/test-utils';
import fake from '__test__/test-utils/constants';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import mockAPI, { generateDefaultMockAPIResponses } from '__test__/test-utils/mockAPI';
import sendInvites from 'utils/data-management/experimentSharing/sendInvites';

import ShareProjectModal from 'components/data-management/project/ShareProjectModal';
import { loadExperiments, removeExperiment } from 'redux/actions/experiments';
import { loadBackendStatus } from 'redux/actions/backendStatus';

jest.mock('@aws-amplify/auth', () => ({
  Auth: {
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
  },
}));

jest.mock('redux/actions/experiments', () => ({
  ...jest.requireActual('redux/actions/experiments'),
  removeExperiment: jest.fn(() => ({ type: 'MOCK_ACTION' })),
}));

jest.mock('utils/data-management/experimentSharing/sendInvites', () => jest.fn(() => Promise.resolve([{
  data: { code: 200 },
},
{
  data: { code: 200 },
}])));

const onCancel = jest.fn();

const experimentId = `${fake.EXPERIMENT_ID}-0`;
const renderShareExperimentModal = async (storeState) => {
  await act(async () => render(
    <Provider store={storeState}>
      <ShareProjectModal
        onCancel={onCancel}
        project={{
          id: experimentId,
          name: fake.EXPERIMENT_NAME,
        }}
        projectType='experiment'
      />
    </Provider>,
  ));
};

describe('Share project modal', () => {
  enableFetchMocks();

  const customAPIResponses = {
    [`/access/${experimentId}$`]: () => Promise.resolve(new Response(JSON.stringify([{
      name: 'Bob',
      email: 'bob@bob.com',
      role: 'explorer',
    },
    {
      name: 'Mocked User',
      email: 'mock@user.name',
      role: 'owner',
    }]))),
  };

  const mockApiResponses = _.merge(
    generateDefaultMockAPIResponses(experimentId), customAPIResponses,
  );

  let storeState = null;

  beforeEach(async () => {
    jest.clearAllMocks();
    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockIf(/.*/, mockAPI(mockApiResponses));

    storeState = makeStore();
    storeState.dispatch(loadBackendStatus(experimentId));
    storeState.dispatch(loadExperiments());
  });

  it('Renders correctly', async () => {
    await renderShareExperimentModal(storeState);

    expect(screen.getByText('Share with collaborators')).toBeInTheDocument();
    expect(screen.getByText(fake.EXPERIMENT_NAME)).toBeInTheDocument();
    expect(screen.getByText('Input an email address. Add multiple addresses with enter.')).toBeInTheDocument();
    expect(screen.getAllByRole('combobox').length).toEqual(2);
    expect(screen.getByText('bob@bob.com')).toBeInTheDocument();
    expect(screen.getByText('Done')).toBeInTheDocument();

    const revokeButtons = screen.getAllByRole('button', { name: 'Revoke' });
    expect(revokeButtons.length).toEqual(2);

    // revoke button for the user viewing is disabled
    expect(revokeButtons[1]).toHaveAttribute('disabled');
  });

  it('Inviting users works', async () => {
    await renderShareExperimentModal(storeState);
    const input = screen.getAllByRole('combobox');
    userEvent.type(input[0], 'asd@asd.com{enter}');

    await waitFor(() => expect(screen.getByText('Add')).toBeInTheDocument());
    const inviteButton = screen.getByText('Add');
    await act(() => fireEvent.click(inviteButton));

    expect(fetchMock.mock.calls.length).toEqual(3);
    expect(fetchMock.mock.calls[2]).toMatchSnapshot();
  });

  it('Revoke access works', async () => {
    await renderShareExperimentModal(storeState);
    const revokeButton = screen.getAllByText('Revoke');
    expect(screen.getByText('bob@bob.com')).toBeInTheDocument();
    await act(() => userEvent.click(revokeButton[0]));
    await waitFor(() => expect(onCancel).toHaveBeenCalled());
  });

  it('Transferring ownership works', async () => {
    await renderShareExperimentModal(storeState);

    // Change role from 'explorer' to 'owner'
    const roleSelect = screen.getAllByRole('combobox')[1];
    userEvent.click(roleSelect);
    const ownerOption = await screen.findByText('Owner');
    userEvent.click(ownerOption);

    // Add a valid email for the new owner
    const emailSelect = screen.getAllByRole('combobox')[0];
    userEvent.type(emailSelect, 'newowner@owner.com{enter}');

    // The OK button now shows "Add" because an email has been added,
    // and since role is 'owner', it is wrapped inside a Popconfirm.
    const transferButton = screen.getByRole('button', { name: 'Add' });
    userEvent.click(transferButton);

    // Wait for the Popconfirm to render and confirm the action.
    const confirmButton = await screen.findByText('Yes');
    userEvent.click(confirmButton);

    await waitFor(() => {
      expect(removeExperiment).toHaveBeenCalledWith(experimentId);
      expect(sendInvites).toHaveBeenCalledWith(['newowner@owner.com'], { id: experimentId, name: 'Test Experiment', role: 'owner' });
    });
  });
});
