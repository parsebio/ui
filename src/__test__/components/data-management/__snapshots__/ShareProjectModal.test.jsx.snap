// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share project modal Inviting users works 1`] = `undefined`;
